packages/react-form-renderer/demo/sandbox.js

```diff
@@ -1,119 +1,26 @@
 /* eslint-disable camelcase */
-import { componentTypes, validatorTypes } from '../src';
-
 const output = {
   title: 'Testing dialog',
   description: 'Description of testing Dialog',
   fields: [
     {
-      name: 'select_1',
+      name: 'dataShower',
       label: 'Test select',
-      component: componentTypes.SELECT,
-      options: [
-        {
-          label: 'Option 1',
-          value: 1
-        },
-        {
-          label: 'Option 2',
-          value: 2
-        }
-      ]
-    },
-    {
-      name: 'text_box_1',
-      label: 'Text Box 1',
-      title: 'Text Box',
-      component: componentTypes.TEXT_FIELD
+      component: 'dataShower',
+      loadOptions: () => new Promise((resolve) => setTimeout(() => resolve({ custom: 'data' }), 1500))
     },
     {
-      name: 'text_box_1_condition',
-      label: 'Text Box conditional',
-      title: 'Text Box',
-      component: componentTypes.TEXT_FIELD,
-      condition: [
-        {
-          when: ['text_box_1', 'text_box_4'],
-          is: 'a'
-        },
-        {
-          when: 'text_box_3',
-          is: 'x'
+      name: 'dataShower',
+      label: 'Test select',
+      component: 'dataShower',
+      actions: {
+        loadOptions: {
+          action: 'loadData',
+          data: { x: 'y' }
```

Comment:
I would call this attribute `args` or `arguments` to make it clear, that it is a function argument. Data might seem misleading based on the action type (function/promise)

Reply:
No, the action will receive the whole object, data is just a name of custom attribute

Reply:
```jsx
action: 'loadData',
superCustomData: { x: 'y' }
```

``jsx
action: 'loadData',
url: 'neco.com'
```

etc.

Reply:
Ah, I thought that it will only accept the data argument. In that, I would go with a slightly different syntax, so we don't force object arguments on functions and people won't have to refactor existing functions:
```jsx
{
  actions: {
    loadOptions: ['loadData' /*name*/, ['foo', 'bar', {x: 'y'}] /*function args*/]
  }
}

// usage
const [name, args] = whereEverTheStuffComesFrom()
loadOptions(...args)
```

Reply:
No, users don't have to refactor anything.

The action mapper looks like this

`loadData: ({ url, query }) => () => origFunc(url, query)`

And I would keep the object notation, as it is consistent with the way how validators are written and also it is definitely more readable than an array.

Reply:
And also, there is no need to do anything in the component itself, function is just replaced by function, as the mapper returns function.

Reply:
I think you missed my point. Forget the load options in the example. What if somebody is using classic `loadOptions(url, query)` arguments they would have create a wrapper around that function to accept arguments form the action mapper. I would like to avoid this:
```jsx
loadOptionsWrapper({url, query}) => loadOptions(url, query)
```

By using the object, we force the object on the action arguments themselves. If you use the spread operator on the argument array, you don't have to do a thing. If you want an object just use an object in the argument array. Keep in mind this action mapper affects every component, not just the select. You can potentially use this to override every single prop of any component.

Reply:
I don't think that creating a wrapper is so huge problem, the advantages that the object notation brings are much more valuable. (And as this is a new feature, there is no refactoring at all...)

Let's say we have an action for translating labels:

Variant a) Object
```jsx
const actionMapper = {
    loadLabel: ({ id }) => intl(id)
}
```

```jsx
{
  name: 'test',
  component: 'text-field',
  actions: {
    label: {
      action: 'loadLabel',
      id: 'cosi'
     }
   }
}
```

With no context, you already know what it does.

Variant b) Array
```jsx
const actionMapper = {
    loadLabel: intl
}
```

It's not clear what the action does, no interface.

```jsx
{
  name: 'test',
  component: 'text-fields',
  actions: {
    label: [ 'loadLabel', 'cosi' ]
   }
}
```

- It's not clear what the array means, what the `cosi` is and with more arguments you need to **keep the order**... 

- No args:

```jsx
label: ['loadLabel']
// or ?
label: 'loadLabel'
```

Then, you have two ways how to do it, or you won't support using just a string and users will be confused as they know this notation from webpack and other configs.

- or let's say we need to skip the first three arguments:

```jsx
label: ['loadLabel', undefined, undefined, undefined, 'cosi']
```

and if you store these values in the database and actions on frontend, then you basically lost all information you need to be able to successfully create the action schema.

Another advantage of the object is that you can validate its schema easily (_are you using loadLabel without id?!_)

Of course, users can simulate the object with passing of object (and they will do as using array is not intuitive), but they can also simulate args array with the same way, if they need:

```
 loadLabel: ({ args }) => intl(...args)

label: {
   action: 'loadLabel',
   args: ['cosi']
}
```

And as users will use object anyway, I think it's better to support good practices. Enforcing is not always bad thing.

Reply:
> I don't think that creating a wrapper is so huge problem, the advantages that the object notation brings are much more valuable

Advantages aside (i do agree here), you are still trying to enforce certain notation of functions. Js functions take an array of arguments and not object.

> And as users will use object anyway, I think it's better to support good practices.

We are not making a linter library. We are not here to decide what is a good/bad practice. If somebody wants to have 10 `undefined` parameters before passing anything valuable, it is up to them. The only thing that we should "enforce" is JS API. And as I said before, functions take an array as an argument, not objects.

Imagine someone crazy that wants to use autogenerated API from OpenAPI. It will generate your functions where you can easily have a ton of undefined values and you have 150 of different callbacks you use in your schemas. Do you really want to make a wrapper around every single one of them?

> It's not clear what the array means, what the cosi is and with more arguments you need to keep the order...

Again I agree, but again you have different tools for adding meaning to attributes like this (JSDoc).

Reply:
> We are not here to decide what is a good/bad practice.

Actually, we are. It's up to us how we define the schema. (Just take a look on validators/ conditions, their notation is basically the same as I propose here. And JavaScript is just current implementation, the schema should not be adjusted because of it.)

> 150 of different callbacks you use in your schemas. Do you really want to make a wrapper around every single one of them?

One-line function. Not great, not terrible.

> (JSDoc).

Using JSDoc (or any of other types of comments) to tell what the code does is not very good idea. Code code should be self-explanatory. (Especially schemas, that can be stored without the code.)

---

But yeah, I will change it.